### DISCUSSION TITLE
How does the canvas re-order elements (nodes/edges) automatically? Can we customize the behaviour?

### DISCUSSION THREAD

**Discussion by Vadorequest** [reaviz/reaflow#55] [Q&A]
When we add new nodes, the canvas redraw itself and somehow "optimize" things.

While it's nice at the beginning, it can also be confusing. For instance, in the below video, the last node added changes the order of the ascendant nodes (inverts them) and this is highly confusing for the end-user.

Is there a way to configure this behaviour anyhow?

Video: https://youtu.be/4JPPXlyoRQ4

**Before** node added:
![image](https://user-images.githubusercontent.com/3807458/107371657-d004df00-6ae4-11eb-8ad2-4ddd735657d9.png)

**After** node added:
![image](https://user-images.githubusercontent.com/3807458/107371725-e6ab3600-6ae4-11eb-99e3-a9e39475daa1.png)

**Comment by amcdnl** [comment]
The order is not really relevant here since its position based -  - which is handled by elkjs.

**Reply by Vadorequest** [reply]
I agree, but moving the whole graph while someone is manipulating it is a big issue for the human brain though.

And, as I plan on making it collaborative, I'd need to avoid the graph to "jump" as soon as somebody adds a new node.

I'll take a look at ELKjs. 👍

**Comment by Vadorequest** [comment]
@amcdnl I think it's related to 

https://github.com/reaviz/reaflow/blob/master/src/layout/elkLayout.ts#L16

But I'm not sure what the other options are, been digging in the ELK documentation and I feel your pain there. 🤯

**Reply by Vadorequest** [reply]
Might be related to http://kieler.github.io/klayjs-d3/examples/interactive/ also, but once again I'm struggling to figure out what the different options are, and I'm not sure if it applies to `org.eclipse.elk.layered` algorithm 🤷

**Reply by Vadorequest** [reply]
I'm trying to figure out where is `nodeLayering` being defined and what are the other values, but I can't find any reference of it, neither in ELKjs nor ELK.

- https://github.com/kieler/elkjs/search?q=nodeLayering&type=code
- https://github.com/eclipse/elk/search?q=nodeLayering

@amcdnl I'm not sure where this property comes from. Is it legit?
